### PR TITLE
sa tokens

### PR DESCRIPTION
**Pull request by deads2k**
sha sha must die!

nee https://github.com/openshift/cluster-kube-apiserver-operator/pull/254 nee https://github.com/openshift/cluster-kube-apiserver-operator/pull/238

### CODE REVIEW

**Comment by deads2k** [comment]
Alright, there's a bug.  Hopefully tomorrow

File: pkg/operator/configobservation/satokencerts/observe_satokencerts.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
+package satokencerts
+
+import (
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
```
Comment (mfojtik) [inline pkg/operator/configobservation/satokencerts/observe_satokencerts.go]: order..

File: pkg/operator/configobservation/satokencerts/observe_satokencerts.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	case len(kcmCerts.Data) > 0:
+		// this means we have this configmap and it has values, so wire up the directory
+		saTokenCertDirs = append(saTokenCertDirs, "/etc/kubernetes/static-pod-resources/configmaps/kube-controller-manager-sa-token-signing-certs")
+	default:
+		// do nothing because aren't going to add a path to a configmap with no files
+	}
+
+	if len(saTokenCertDirs) > 0 {
+		if err := unstructured.SetNestedStringSlice(observedConfig, saTokenCertDirs, saTokenCertsPath...); err != nil {
+			errs = append(errs, err)
```
Comment (mfojtik) [inline pkg/operator/configobservation/satokencerts/observe_satokencerts.go]: should we return here?

Comment (mfojtik) [reply pkg/operator/configobservation/satokencerts/observe_satokencerts.go]: nvmd.

**Comment by mfojtik** [comment]
pppanic in operator pod: `connect: connection refused`

File: pkg/operator/configobservation/satokencerts/observe_satokencerts.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+// There are two possible sources: the installer and the kube-controller-manger-operator, but we wire to the target namespace
+// to avoid setting a config we cannot fulfill which would crash the kube-apiserver.
+func ObserveSATokenCerts(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	errs := []error{}
+	prevObservedConfig := map[string]interface{}{}
+
+	// copy non-empty .serviceAccountPublicKeyFiles from existingConfig to prevObservedConfig
+	saTokenCertsPath := []string{"serviceAccountPublicKeyFiles"}
+	existingSATokenCerts, _, err := unstructured.NestedStringSlice(existingConfig, saTokenCertsPath...)
```
Comment (mfojtik) [inline pkg/operator/configobservation/satokencerts/observe_satokencerts.go]: can you add `found` here and then check at line 30? (`if !found || len(...)`)

File: pkg/operator/configobservation/satokencerts/observe_satokencerts.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		// do nothing because aren't going to add a path to a configmap with no files
+	}
+
+	if len(saTokenCertDirs) > 0 {
+		if err := unstructured.SetNestedStringSlice(observedConfig, saTokenCertDirs, saTokenCertsPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if !equality.Semantic.DeepEqual(existingSATokenCerts, saTokenCertDirs) {
```
Comment (mfojtik) [inline pkg/operator/configobservation/satokencerts/observe_satokencerts.go]: `&& len(errs) == 0`

File: pkg/operator/targetconfigcontroller/targetconfigcontroller.go [hunk 5]
```diff
@@ -103,10 +108,27 @@ func (c TargetConfigController) sync() error {
...
+	if err := json.NewDecoder(bytes.NewBuffer(operatorConfig.Spec.ObservedConfig.Raw)).Decode(&existingConfig); err != nil {
+		glog.V(4).Infof("decode of existing config failed with error: %v", err)
+	}
+	requiredPaths := [][]string{
+		{"serviceAccountPublicKeyFiles"},
+	}
+	for _, requiredPath := range requiredPaths {
+		_, found, _ := unstructured.NestedFieldNoCopy(existingConfig, requiredPath...)
+		if !found {
+			message := fmt.Sprintf("%v missing from config", strings.Join(requiredPath, "."))
```
Comment (mfojtik) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller.go]: nit: you already have `fmt` imported, how about:

```go
err := fmt.Errorf("...")
c.eventRecorder.Warning("ConfigMissing", message.Error())
return err
```

File: pkg/operator/configobservation/satokencerts/observe_satokencerts.go (hunk 3 above)
Comment (deads2k) [reply pkg/operator/configobservation/satokencerts/observe_satokencerts.go]: > can you add `found` here and then check at line 30? (`if !found || len(...)`)

I thing missing should be missing

**Pull request merged**